plugins/browser-plugin-form-tracking/src/helpers.ts

```diff
@@ -78,237 +78,351 @@ interface FormConfiguration {
-            ) {
-              if (config.eventFilter(FormTrackingEvent.FOCUS_FORM)) {
-                addEventListener(
-                  innerElement,
-                  'focus',
-                  getFormChangeListener(tracker, config, 'focus_form', context),
-                  false
-                );
-              }
-              if (config.eventFilter(FormTrackingEvent.CHANGE_FORM)) {
-                addEventListener(
-                  innerElement,
-                  'change',
-                  getFormChangeListener(tracker, config, 'change_form', context),
-                  false
-                );
-              }
-              innerElement[trackingMarker] = true;
-            }
-          });
-      });
+  const events = options?.events ?? defaultFormTrackingEvents;
 
-      if (!form[trackingMarker]) {
-        if (config.eventFilter(FormTrackingEvent.SUBMIT_FORM)) {
-          addEventListener(form, 'submit', getFormSubmissionListener(tracker, config, trackingMarker, context));
-        }
-        form[trackingMarker] = true;
-      }
-    }
+  const targets = (_targets[tracker.id] = getTargetList(options?.targets, config.forms));
+
+  if (events.indexOf(FormTrackingEvent.FOCUS_FORM) !== -1) {
+    _focusListeners[tracker.id] = getFormChangeListener(tracker, config, FormTrackingEvent.FOCUS_FORM, context);
+    targets.forEach((target) => addEventListener(target, 'focus', _focusListeners[tracker.id], true));
+  }
+  if (events.indexOf(FormTrackingEvent.CHANGE_FORM) !== -1) {
+    _changeListeners[tracker.id] = getFormChangeListener(tracker, config, FormTrackingEvent.CHANGE_FORM, context);
+    targets.forEach((target) => addEventListener(target, 'change', _changeListeners[tracker.id], true));
+  }
+  if (events.indexOf(FormTrackingEvent.SUBMIT_FORM) !== -1) {
+    _submitListeners[tracker.id] = getFormSubmissionListener(tracker, config, context);
+    targets.forEach((target) => addEventListener(target, 'submit', _submitListeners[tracker.id], true));
+  }
+}
+
+/**
+ * Builds a list of targets for the plugin event listeners
+ *
+ * The list can include any specifically provided targets, and will be extended to include the root nodes of any explicit HTMLFormElements provided
+ * With neither provided, defaults to the current page's `document` element
+ *
+ * @param configTargets Explicitly configured list of event target listeners, if any
+ * @param forms Explicitly configured list of form elements to track, if any
+ * @returns List of EventTargets to add the listener to
+ */
+function getTargetList(configTargets: EventTarget[] | undefined, forms: FormConfiguration['forms']) {
+  // we attach to document rather than window because the window focus event occurs more often than we require
+  const targets = configTargets ?? [document];
+
+  if (forms) {
+    Array.prototype.forEach.call(forms, (form: HTMLFormElement) => {
+      targets.push(form.ownerDocument.documentElement);
```

Comment:
I am probably missing something, but is it possible that if two forms were passed here that come from the same document, we will subscribe to the form events from that document twice (will track duplicate form event)?

Reply:
We'll call addEventListener multiple times, yes, but because we use an identical listener function instance they will naturally dedupe and only fire once.

https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener

> If the function or object is already in the list of event listeners for this target, the function or object is not added a second time.

Though we can probably do a quick uniqueness pass first if you think it'd be safer?

Reply:
Ahh, wasn't aware that it dedupes by the listener function! That sounds good!